lib/deb_utils.rb

```diff
@@ -0,0 +1,43 @@
+def extract_deb(file, target = nil)
+  # extract_deb: unarchive .deb files
+  # Usage: extract_deb(<file>, <target (optional)>)
+  #   file: a .deb archive file
+  #   target: (optional) a specified file to extract from archive, should be passed in regex
+  #
+  # Example:
+  #   extract_deb('example.deb', /data\..*/) # extract files from example.deb with filenames matching the /data\..*/ regex
+  #   extract_deb('example.deb')             # extract all files in example.deb
+  #
+  file_found = false if target
+  src_fileIO = File.open(file, 'rb')
+  # get first line of the given file, should be a signature string (`!<arch>\n`) if it is a valid deb file
+  signature = src_fileIO.gets
+
+  abort 'Malformed archive :/' unless signature == "!<arch>\n"
+
+  # process each file in archive
+  while (line = src_fileIO.gets) do
+    # read file meta
+    name, modtime, uid, gid, mode, size, end_char = \
+      line.scan(/(.{16})(.{12})(.{6})(.{6})(.{8})(.{10})(.{1})/).flatten.map(&:strip)
+
+    # check ending byte
+    abort "Malformed archive :/" unless end_char == '`'
```

Comment:
This is good enough to determine if an archive is malformed?

Reply:
An `ar` archive is formed by readable texts only (except data stored in the archive, like `.gz` files), the ``` symbol will added to the end of every file metadata line, see https://en.wikipedia.org/wiki/Ar_(Unix)